bgpd/bgp_mplsvpn.c

```diff
@@ -1350,6 +1349,187 @@ leak_update(struct bgp *to_bgp, struct bgp_dest *bn,
 	return new;
 }
 
+static void
+bgp_mplsvpn_path_nh_label_map(struct bgp_path_info *path,
+			      struct bgp_label_per_nexthop_cache *blnc,
+			      bool make)
+{
+	if (path->label_nexthop) {
+		LIST_REMOVE(path, label_nh_thread);
+		path->label_nexthop->path_count--;
+		path->label_nexthop = NULL;
+	}
+	if (make) {
+		LIST_INSERT_HEAD(&(blnc->paths), path, label_nh_thread);
+		path->label_nexthop = blnc;
+		path->label_nexthop->path_count++;
+	}
+	if (blnc)
+		blnc->last_update = monotime(NULL);
+}
+
+void bgp_mplsvpn_path_nh_label_unlink(struct bgp_path_info *path)
+{
+	struct bgp_label_per_nexthop_cache *blnc = path->label_nexthop;
+
+	if (!blnc)
+		return;
+
+	bgp_mplsvpn_path_nh_label_map(path, NULL, false);
+	if (LIST_EMPTY(&(blnc->paths)))
+		bgp_label_per_nexthop_free(blnc);
+}
+static void
+bgp_mplsvpn_path_nh_label_update(struct bgp_label_per_nexthop_cache *blnc)
+{
+	struct bgp_table *table;
+	struct bgp_dest *dest;
+	struct bgp *bgp;
+	struct bgp_path_info *path;
+
+	LIST_FOREACH (path, &(blnc->paths), label_nh_thread) {
+		dest = path->net;
+		if (!dest)
+			continue;
+		table = bgp_dest_table(dest);
+		if (!table)
+			continue;
+		bgp = table->bgp;
+		vpn_leak_from_vrf_update(blnc->to_bgp, bgp, path);
+	}
+}
+
+static int bgp_mplsvpn_get_label_cb(mpls_label_t label, void *context,
+				    bool allocated)
+{
+	struct bgp_label_per_nexthop_cache *blnc =
+		(struct bgp_label_per_nexthop_cache *)context;
+	mpls_label_t old_label = blnc->label;
+	int debug = BGP_DEBUG(vpn, VPN_LEAK_LABEL);
+	struct nexthop *nhop;
+
+	if (debug)
+		zlog_debug("%s: label=%u, allocated=%d, prefix=%pFX", __func__,
+			   label, allocated, &blnc->prefix);
+	if (!allocated)
+		/*
+		 * previously-allocated label is now invalid
+		 * eg: zebra deallocated the labels and notifies it
+		 */
+		blnc->label = MPLS_INVALID_LABEL;
+	else
+		/* update the entry with the new label */
+		blnc->label = label;
+	/* update paths */
+	if (old_label == blnc->label)
+		return 0;
+
+	if (blnc->label != MPLS_INVALID_LABEL) {
+		for (nhop = blnc->nh_list; nhop; nhop = nhop->next) {
+			bgp_zebra_send_nexthop_label(
+				ZEBRA_MPLS_LABELS_ADD, blnc->label,
+				nhop->ifindex, nhop->vrf_id, ZEBRA_LSP_BGP,
+				&blnc->prefix);
+		}
+	}
+
+	bgp_mplsvpn_path_nh_label_update(blnc);
+
+	return 0;
+}
+
+/* check if the path is eligible for mpls allocation per next-hop
+ * return true:
+ * - while there is no valid bnc
+ * - if there is only nexthop ifindex bnc nexthop
+ * return false:
+ * if true, then try to return available label if any
+ */
+static bool bgp_mplsvpn_get_label(struct bgp_path_info *pi, struct bgp *to_bgp,
+				  struct bgp *from_bgp, mpls_label_t *label)
+{
+	struct bgp_nexthop_cache *bnc = pi->nexthop;
+	struct prefix *p;
+	struct nexthop *nhop, *iter;
+	struct bgp_label_per_nexthop_cache *blnc;
+	struct bgp_label_per_nexthop_cache_head *tree;
+	bool found_other_nexthop = false;
+	bool ret = true;
```

Comment:
This variable is not used as it should be (not assigned a value), is it really worth having it?

Reply:
we have two failing cases:
- the BGP update is not eligible for using a label for a specific nexthop
- the BGP update is eligible for using a label for a specific nexthop, but we do not have yet a label
In the latter case, I just want to wait the new label being offered.

Reply:
ok